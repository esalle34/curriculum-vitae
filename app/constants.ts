enum Pages {
  CV = "cv",
  COMMON = "common",
}

export { Pages };
