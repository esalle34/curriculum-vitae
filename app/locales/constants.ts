enum Languages {
  English = "en",
  Français = "fr",
  "Español" = "es",
}

enum Countries {
  France = "FR",
  USA = "US",
  "United Kingdom" = "GB",
  "España" = "ES",
  Luxembourg = "LU",
  Suisse = "CH",
  Belgique = "BE",
}

const languagesArray = Object.entries(Languages);
const countriesArray = Object.entries(Countries);

const locales = countriesArray.flatMap(([_countryKey, country]) =>
  languagesArray.map(
    ([_languageKey, language]): Locale => `${language}-${country}`
  )
);

export type Locale = `${Languages}-${Countries}`;
const defaultLocale: Locale = `${Languages.English}-${Countries.USA}`;

interface DecodeTranslationOptions {
  className?: string;
  htmlTag: string;
}

export interface DecodeTranslation {
  options: DecodeTranslationOptions;
  translation: string;
}

export {
  Languages,
  Countries,
  defaultLocale,
  languagesArray,
  countriesArray,
  locales,
};
