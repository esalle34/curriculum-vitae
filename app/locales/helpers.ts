import { Countries, DecodeTranslation, Languages, Locale } from "./constants";

const getDictionary = (language: Languages, page: string) => ({
  [language]: () =>
    import(`@/app/locales/${language}/${page}.json`).then(
      (module) => module.default
    ),
});

const getLanguageAndCountry = (locale: Locale) => {
  const [language, country] = locale.split("-");

  return { language: language as Languages, country: country as Countries };
};

const decodeTranslation = ({
  options: { htmlTag, className },
  translation,
}: DecodeTranslation) => {
  return translation
    .replace(`<0>`, `<${htmlTag} ${className ? `class="${className}"` : ""}>`)
    .replace(`</0>`, `</${htmlTag}>`);
};

export { getDictionary, getLanguageAndCountry, decodeTranslation };
