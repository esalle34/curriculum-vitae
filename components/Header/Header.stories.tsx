import type { Meta, StoryObj } from "@storybook/react";

import Header from "./Header";
import { Languages, Countries } from "@/app/locales/constants";

const meta: Meta<typeof Header> = {
  component: Header,
};

export default meta;

type Story = StoryObj<typeof Header>;

/*
 *ðŸ‘‡ Render functions are a framework specific feature to allow you control on how the component renders.
 * See https://storybook.js.org/docs/api/csf
 * to learn how to use render functions.
 */
export const HeaderFR: Story = {
  render: () => <Header locale={`${Languages.FranÃ§ais}-${Countries.France}`} />,
};
