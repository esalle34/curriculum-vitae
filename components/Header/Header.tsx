import { Pages } from "@/app/constants";
import { Locale } from "@/app/locales/constants";
import { getLanguageAndCountry, getDictionary } from "@/app/locales/helpers";
import HeaderNav from "../HeaderNav";

interface Props {
  locale: Locale;
}

const Header = async ({ locale }: Props) => {
  const { language, country } = getLanguageAndCountry(locale);

  const dict = await getDictionary(language, Pages.COMMON)[language]();

  return (
    <header className="bg-blue-800">
      <HeaderNav language={language} country={country} dict={dict} />
    </header>
  );
};

export default Header;
