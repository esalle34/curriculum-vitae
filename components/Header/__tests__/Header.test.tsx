import { Countries, Languages } from "@/app/locales/constants";
import Header from "..";
import { render } from "@testing-library/react";

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
}));

describe("Header", () => {
  it("should render FR snapshot correctly", async () => {
    const Result = await Header({
      locale: `${Languages.FranÃ§ais}-${Countries.France}`,
    });
    const { container } = render(Result);
    expect(container).toMatchSnapshot();
  });
  it("should render EN snapshot correctly", async () => {
    const Result = await Header({
      locale: `${Languages.English}-${Countries.USA}`,
    });
    const { container } = render(Result);
    expect(container).toMatchSnapshot();
  });
});
