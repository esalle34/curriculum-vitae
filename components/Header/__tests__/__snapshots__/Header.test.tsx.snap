// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render EN snapshot correctly 1`] = `
<div>
  <header
    class="bg-sky-800 w-full"
  >
    <nav
      class="flex"
    >
      <div
        class="flex w-full justify-between items-center"
      >
        <a
          href="https://www.linkedin.com/in/%F0%9F%92%BB-eric-salle-431664109/"
        >
          <h1
            class="pl-4 text-gray-100 font-bold text-2xl my-2 font-instagram"
          >
            Eric Salle
          </h1>
        </a>
        <button
          class="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-sky-800"
        >
          EN
           - 
          🇺🇸
        </button>
      </div>
      <div
        class="hidden"
        data-testid="popin"
      >
        <div
          class="flex flex-col gap-2 md:w-1/4 bg-slate-100 py-4 px-6 rounded z-10 border-2 border-slate-200"
        >
          <div
            class="flex justify-between items-center"
          >
            <h3
              class="py-2 text-lg font-semibold text-sky-800"
            >
              Language & Country
            </h3>
            <img
              alt="Close"
              class="cursor-pointer"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              role="img"
              src="/images/cross.svg"
              style="color: transparent;"
              title="Close"
              width="20"
            />
          </div>
          <select
            class="my-2 p-2 border-2 border-slate-300 rounded"
          >
            <option
              value="FR"
            >
              France
               - 
              🇫🇷
            </option>
            <option
              selected=""
              value="US"
            >
              USA
               - 
              🇺🇸
            </option>
            <option
              value="GB"
            >
              United Kingdom
               - 
              🇬🇧
            </option>
            <option
              value="ES"
            >
              España
               - 
              🇪🇸
            </option>
            <option
              value="LU"
            >
              Luxembourg
               - 
              🇱🇺
            </option>
            <option
              value="CH"
            >
              Suisse
               - 
              🇨🇭
            </option>
            <option
              value="BE"
            >
              Belgique
               - 
              🇧🇪
            </option>
          </select>
          <select
            class="my-2 p-2 border-2 border-slate-300 rounded"
          >
            <option
              selected=""
              value="en"
            >
              English
            </option>
            <option
              value="fr"
            >
              Français
            </option>
            <option
              value="es"
            >
              Español
            </option>
          </select>
          <button
            class="self-end rounded-lg text-sky-800 border-sky-800 hover:bg-sky-800 hover:text-white border-2 p-2"
          >
            Submit
          </button>
        </div>
        <div
          class="hidden"
          data-testid="background"
        />
      </div>
    </nav>
  </header>
</div>
`;

exports[`Header should render FR snapshot correctly 1`] = `
<div>
  <header
    class="bg-sky-800 w-full"
  >
    <nav
      class="flex"
    >
      <div
        class="flex w-full justify-between items-center"
      >
        <a
          href="https://www.linkedin.com/in/%F0%9F%92%BB-eric-salle-431664109/"
        >
          <h1
            class="pl-4 text-gray-100 font-bold text-2xl my-2 font-instagram"
          >
            Eric Salle
          </h1>
        </a>
        <button
          class="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-sky-800"
        >
          FR
           - 
          🇫🇷
        </button>
      </div>
      <div
        class="hidden"
        data-testid="popin"
      >
        <div
          class="flex flex-col gap-2 md:w-1/4 bg-slate-100 py-4 px-6 rounded z-10 border-2 border-slate-200"
        >
          <div
            class="flex justify-between items-center"
          >
            <h3
              class="py-2 text-lg font-semibold text-sky-800"
            >
              Langue & Pays
            </h3>
            <img
              alt="Fermer"
              class="cursor-pointer"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              role="img"
              src="/images/cross.svg"
              style="color: transparent;"
              title="Fermer"
              width="20"
            />
          </div>
          <select
            class="my-2 p-2 border-2 border-slate-300 rounded"
          >
            <option
              selected=""
              value="FR"
            >
              France
               - 
              🇫🇷
            </option>
            <option
              value="US"
            >
              USA
               - 
              🇺🇸
            </option>
            <option
              value="GB"
            >
              United Kingdom
               - 
              🇬🇧
            </option>
            <option
              value="ES"
            >
              España
               - 
              🇪🇸
            </option>
            <option
              value="LU"
            >
              Luxembourg
               - 
              🇱🇺
            </option>
            <option
              value="CH"
            >
              Suisse
               - 
              🇨🇭
            </option>
            <option
              value="BE"
            >
              Belgique
               - 
              🇧🇪
            </option>
          </select>
          <select
            class="my-2 p-2 border-2 border-slate-300 rounded"
          >
            <option
              value="en"
            >
              English
            </option>
            <option
              selected=""
              value="fr"
            >
              Français
            </option>
            <option
              value="es"
            >
              Español
            </option>
          </select>
          <button
            class="self-end rounded-lg text-sky-800 border-sky-800 hover:bg-sky-800 hover:text-white border-2 p-2"
          >
            Valider
          </button>
        </div>
        <div
          class="hidden"
          data-testid="background"
        />
      </div>
    </nav>
  </header>
</div>
`;
